Simulator has global dependence that prevents repetition from working consistently

This do-nothing test:

```
def test_repetition():
    engine_list = []
    for _ in range(2):
        backend = Simulator()
        eng = MainEngine(backend=backend, engine_list=engine_list)
        eng.allocate_qureg(1)
        _, state = backend.cheat()
```

Fails with this error:

```
self = <projectq.backends._sim._pysim.Simulator object at 0x7ff7ff10b5d0>, ID = 0

    def allocate_qubit(self, ID):
        """
            Allocate a qubit.
    
            Args:
                ID (int): ID of the qubit which is being allocated.
            """
        self._map[ID] = self._num_qubits
        self._num_qubits += 1
>       self._state.resize(1 << self._num_qubits)
E       ValueError: cannot resize an array that references or is referenced
E       by another array in this way.  Use the resize function

projectq/backends/_sim/_pysim.py:112: ValueError
```

I think it has something to do with cache invalidation, because if I use a new empty-engine list for each call then the test runs to completion.

--- a/projectq/cengines/_main.py
+++ b/projectq/cengines/_main.py
@@ -110,7 +110,7 @@ class MainEngine(BasicEngine):
                         "E.g. MainEngine(engine_list=[AutoReplacer]) instead "
                         "of\n     MainEngine(engine_list=[AutoReplacer()])")
 
-        engine_list.append(backend)
+        engine_list = engine_list + [backend]
         self.backend = backend
 
         # Test that user did not supply twice the same engine instance

--- a/projectq/cengines/_main_test.py
+++ b/projectq/cengines/_main_test.py
@@ -24,7 +24,8 @@ def test_main_engine_init():
     ceng1 = DummyEngine()
     ceng2 = DummyEngine()
     test_backend = DummyEngine()
-    eng = _main.MainEngine(backend=test_backend, engine_list=[ceng1, ceng2])
+    engine_list = [ceng1, ceng2]
+    eng = _main.MainEngine(backend=test_backend, engine_list=engine_list)
     assert id(eng.next_engine) == id(ceng1)
     assert id(eng.main_engine) == id(eng)
     assert not eng.is_last_engine
@@ -37,6 +38,7 @@ def test_main_engine_init():
     assert test_backend.is_last_engine
     assert id(test_backend.main_engine) == id(eng)
     assert not test_backend.next_engine
+    assert len(engine_list) == 2
 
 
 def test_main_engine_init_failure():

